backend/src/connector/mod.rs

```diff
@@ -1 +1,328 @@
+    PermissionDenied,
+    #[error("no client found for requested resource")]
+    NoClient,
+    #[error("can't form hyper request")]
+    CantFormRequest,
+    #[error("bad URI")]
+    BadUri,
+}
+
+/// HTTP Connector constructor
+#[derive(Debug)]
+pub struct Connector;
+
+impl Connector {
+    /// Alows building a HTTP(S) connector. Used for instantiating clients with custom
+    /// connectors.
+    #[must_use]
+    pub const fn builder() -> Builder {
+        Builder {}
+    }
+}
+
+/// Builder for HTTP(S) connectors
+#[derive(Debug)]
+pub struct Builder {}
+
+impl Builder {
+    /// [Stub] Use HTTPS instead of HTTP
+    #[must_use]
+    pub const fn https(self) -> HttpsBuilder {
+        HttpsBuilder {}
+    }
+
+    /// Build a HTTP connector
+    #[must_use]
+    pub fn build(self) -> hyper::client::connect::HttpConnector {
+        hyper::client::connect::HttpConnector::new()
+    }
+}
+
+// TODO
+/// [Stub] Builder for HTTPS connectors
+#[derive(Debug)]
+pub struct HttpsBuilder {}
+
+impl HttpsBuilder {
+    pub fn build(self) {
+        unimplemented!()
+    }
+}
+
+#[derive(Clone)]
+pub struct BobClient<Client: ApiNoContext<ClientContext> + Send + Sync> {
+    /// Unique Identifier
+    id: Uuid,
+
+    /// Bob's hostname
+    hostname: Hostname,
+
+    // NOTE: Can (and should) the API interface mutate?..
+    /// Connection,
+    main: Arc<Client>,
+
+    /// Clients for all known nodes
+    cluster: HashMap<NodeName, Arc<Client>>,
+}
+
+#[allow(clippy::missing_fields_in_debug)]
+impl<Client: ApiNoContext<ClientContext> + Send + Sync + Clone> std::fmt::Debug
+    for BobClient<Client>
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let user = &self
+            .main
+            .context()
+            .auth_data
+            .as_ref()
+            .map_or("Unknown", |cred| cred.username());
+        f.debug_struct("BobClient")
+            .field("hostname", &self.hostname)
+            .field("user", &user)
+            .finish()
+    }
+}
+
+impl<ApiInterface: ApiNoContext<ClientContext> + Send + Sync> BobClient<ApiInterface> {
+    // impl BobClient<C> {
+    /// Creates new [`BobClient`] from [`BobConnectionData`]
+    ///
+    /// # Errors
+    /// The function will fail if a hostname isn't a valid url or the client couldn't establish
+    /// connection for the BOB cluster
+    pub async fn try_new(
+        bob_data: BobConnectionData,
+        timeout: RequestTimeout,
+    ) -> Result<HttpBobClient, ClientError> {
+        let auth = bob_data
+            .credentials
+            .map(|creds| Authorization::basic(&creds.login, &creds.password));
+        let hostname = bob_data.hostname.clone();
+
+        let context: ClientContext = ClientContext {
+            timeout,
+            auth_data: auth,
+            xspan: XSpanIdString::default(),
+        };
+        let client =
+            Client::try_new_http(&hostname.to_string()).change_context(ClientError::InitClient)?;
```

Comment:
Please, add context to the error with `hostname` inside. This will make it easier to diagnose problems